Source/Tools/LinuxEmulation/LinuxSyscalls/Seccomp/BPFEmitter.cpp

```diff
@@ -0,0 +1,378 @@
+
+    switch (Op) {
+    case BPF_ADD: EMIT_INST(add(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_SUB: EMIT_INST(sub(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_MUL: EMIT_INST(mul(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_DIV:
+      // Specifically unsigned.
+      EMIT_INST(udiv(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg));
+      break;
+    case BPF_OR: EMIT_INST(orr(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_AND: EMIT_INST(and_(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_LSH: EMIT_INST(lslv(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_RSH: EMIT_INST(lsrv(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    case BPF_MOD:
+      // Specifically unsigned.
+      EMIT_INST(udiv(ARMEmitter::Size::i32Bit, REG_TMP2, REG_A, SrcReg));
+      EMIT_INST(msub(ARMEmitter::Size::i32Bit, REG_A, REG_TMP2, SrcReg, REG_A));
+      break;
+    case BPF_XOR: EMIT_INST(eor(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg)); break;
+    default: RETURN_ERROR(-EINVAL);
+    }
+
+    break;
+  }
+  case BPF_NEG:
+    // Only BPF_K supported on NEG.
+    VALIDATE(SrcType == BPF_K);
+
+    EMIT_INST(neg(ARMEmitter::Size::i32Bit, REG_A, REG_A));
+    break;
+
+  default: RETURN_ERROR(-EINVAL);
+  }
+
+  RETURN_SUCCESS();
+}
+
+template<bool CalculateSize>
+uint64_t BPFEmitter::HandleJmp(uint32_t BPFIP, uint32_t NumInst, const sock_filter* Inst) {
+  [[maybe_unused]] size_t OpSize {};
+  const auto SrcType = BPF_SRC(Inst->code);
+  const auto Op = BPF_OP(Inst->code);
+
+  switch (Op) {
+  case BPF_JA: {
+    // Only BPF_K supported on JA.
+    VALIDATE(SrcType == BPF_K);
+
+    // BPF IP register is effectively only 32-bit. Treat k constant like a signed integer.
+    // This allows it to jump anywhere in the program.
+    // But! Loops are EXPLICITLY disallowed inside of BPF programs.
+    // This is to prevent DOS style attacks through BPF programs.
+    uint64_t Target = BPFIP + Inst->k + 1;
+    // Jumped past the end.
+    VALIDATE(Target < NumInst);
+
+    fextl::unordered_map<uint32_t, ARMEmitter::ForwardLabel>::iterator TargetLabel {};
+
+    if constexpr (!CalculateSize) {
+      TargetLabel = JumpLabels.try_emplace(Target, ARMEmitter::ForwardLabel {}).first;
+    }
+
+    EMIT_INST(b(&TargetLabel->second));
+    break;
+  }
+  case BPF_JEQ:
+  case BPF_JGT:
+  case BPF_JGE:
+  case BPF_JSET: {
+    auto CompareSrcReg = REG_X;
+    if (SrcType == BPF_K) {
+      CompareSrcReg = REG_TMP;
+      auto Const = ConstPool.try_emplace(Inst->k, ARMEmitter::ForwardLabel {});
+      EMIT_INST(ldr(CompareSrcReg, &Const.first->second));
+    }
+    uint32_t TargetTrue = BPFIP + Inst->jt + 1;
+    uint32_t TargetFalse = BPFIP + Inst->jf + 1;
+
+    // Trying to jump past the end.
+    VALIDATE(TargetTrue < NumInst && TargetFalse < NumInst);
+
+    auto CompareResultOp = ARMEmitter::Condition::CC_EQ;
```

Comment:
Default value is confusing here, how about just `ARMEmitter::Condition CompareResultOp;`

Reply:
Default value removed.